src/shogun/io/streaming/ParseBuffer.h

```diff
@@ -99,11 +99,14 @@ template <class T> class CParseBuffer: public CSGObject
 	Example<T>* get_free_example()
 	{
 #ifdef HAVE_CXX11
-		std::unique_lock<std::mutex> write_lk(*write_mutex, std::defer_lock);
-		std::unique_lock<std::mutex> current_ex_lock(*ex_in_use_mutex[ex_write_index], std::defer_lock);
-		std::lock(write_lk, current_ex_lock);
-		while (ex_used[ex_write_index] == E_NOT_USED)
-			ex_in_use_cond[ex_write_index]->wait(current_ex_lock);
+//		std::unique_lock<std::mutex> write_lk(*write_mutex, std::defer_lock);
+//		std::unique_lock<std::mutex> current_ex_lock(*ex_in_use_mutex[ex_write_index], std::defer_lock);
+//		std::lock(write_lk, current_ex_lock);
+        int32_t current_write_index = ex_write_index.load(std::memory_order_relaxed);  // only written from parser thread
+        int32_t current_read_index = ex_read_index.load(std::memory_order_acquire); // for synchronisation
+		while (ex_used[current_write_index] == E_NOT_USED)
```

Comment:
```C++
while (ex_used[current_write_index] == E_NOT_USED)
         current_read_index = ex_read_index.load(std::memory_order_acquire); // for synchronisation
```
This is what I think:
Even though the `ex_used` itself is not atomic, doing `ex_read_index.load(std::memory_order_acquire);`(soon will replace it by `std::atomic_thread_fence(std::memory_order_acquire);` ) in every iteration will synchronize every store before `ex_read_index.store( ... ,std::memory_order_release);` even the  non atomic ones.(as suggested [here](https://gcc.gnu.org/wiki/Atomic/GCCMM/AtomicSync) ) So changes to the `ex_used` will be reflected after the `std::memory_order_acquire`. 

Let me know what you think.